torchbenchmark/models/sam/__init__.py

```diff
@@ -47,14 +84,72 @@ def get_module(self):
         return self.model, (example_input, multimask_output)
             
     def train(self):
-        error_msg = """
-            As of May 17, 2023
-            Some base VIT checkpoints are available for SAM but getting the dataset
-            requires a research license. It's easy to make up a training loop on random
-            data and if that's interesting please let @msaroufim know
-            https://github.com/facebookresearch/segment-anything#dataset
-        """
-        return NotImplementedError(error_msg)
+        # Code copied from https://github.com/bowang-lab/MedSAM/blob/main/finetune_and_inference_tutorial.py
+        
+        
+        npz_tr_path = 'data/Npz_files/CT_Abd-Gallbladder/train'
+        work_dir = './work_dir'
+        task_name = 'CT_Abd-Gallbladder'
+        # prepare SAM model
+        model_type = 'vit_b'
+        checkpoint = 'work_dir/SAM/sam_vit_b_01ec64.pth'
+        device = 'cuda:0'
+        model_save_path = os.path.join(work_dir, task_name)
+        os.makedirs(model_save_path, exist_ok=True)
+        self.model.train()
+        optimizer = torch.optim.Adam(self.model.mask_decoder.parameters(), lr=1e-5, weight_decay=0)
+        seg_loss = monai.losses.DiceCELoss(sigmoid=True, squared_pred=True, reduction='mean')
+
+        num_epochs = 100
+        losses = []
+        best_loss = 1e10
+
+        train_dataset = NpzDataset(npz_tr_path)
+        train_dataloader = DataLoader(train_dataset, batch_size=64, shuffle=True)
+        for epoch in range(num_epochs):
+            epoch_loss = 0
+            # train
+            for step, (image_embedding, gt2D, boxes) in enumerate(tqdm(train_dataloader)):
+                # do not compute gradients for image encoder and prompt encoder
+                with torch.no_grad():
```

Comment:
It is suggested to use the new `staged train` interface if possible, users implement 3 interfaces for train: `forward()`, `backward()`, and `optimizer_step()`,

Reply:
You mean the same interface that the optimizer tests are using in the e2e benchmark folder?

Reply:
@msaroufim Not really, I am talking about these interfaces in core models: https://github.com/pytorch/benchmark/blob/main/torchbenchmark/util/model.py#L251-L258